### PR TITLE
Test/repo privatisation

### PR DESCRIPTION
**Pull request by alexanderleegs**
Adds tests for repo privatisation.

### CODE REVIEW

File: src/services/configServices/__tests__/SettingsService.spec.js
Comment (seaerchin) [inline src/services/configServices/__tests__/SettingsService.spec.js:1]: tests for adding/updating pw added in prev pr right

File: src/routes/v2/authenticatedSites/settings.js [hunk 1]
```diff
@@ -129,7 +130,7 @@ class SettingsRouter {
     )
     router.post(
       "/repo-password",
-      attachRollbackRouteHandlerWrapper(this.updateRepoPassword)
+      attachWriteRouteHandlerWrapper(this.updateRepoPassword)
```
Comment (seaerchin) [inline src/routes/v2/authenticatedSites/settings.js:133]: sorry, could you explain why this change was amde?

File: src/routes/v2/authenticatedSites/__tests__/Settings.spec.js [hunk 2]
```diff
@@ -61,6 +74,9 @@ describe("Settings Router", () => {
     sha: homepageSha,
   }
 
+  const PASSWORD = "password"
```
Comment (seaerchin) [inline src/routes/v2/authenticatedSites/__tests__/Settings.spec.js:77]:
```suggestion
  const MOCK_PASSWORD = "password"
```

File: src/integration/Privatisation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+  afterAll(async () => {
+    await User.sync({ force: true })
+    await Site.sync({ force: true })
+    await Repo.sync({ force: true })
+    await SiteMember.sync({ force: true })
+    await Deployment.sync({ force: true })
+  })
+
+  describe("GET /repo-password", () => {
+    it("should successfully retrieve password for an amplify site", async () => {
```
Comment (seaerchin) [inline src/integration/Privatisation.spec.ts:264]:
```suggestion
    it("should successfully retrieve password for an amplify site with a password", async () => {
```

File: src/integration/Privatisation.spec.ts [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+      // Arrange
+      const app = generateRouterForUserWithSite(
+        subrouter,
+        MOCK_USER_SESSION_DATA_TWO,
+        MOCK_REPO_NAME_TWO
```
Comment (seaerchin) [inline src/integration/Privatisation.spec.ts:269]: can we alias this for clarity in the tests?

ie, 
```ts
const MOCK_AMPLIFY_REPO_WITH_PASSWORD = MOCK_REPO_NAME_TWO
```

File: src/integration/Privatisation.spec.ts [hunk 5]
```diff
@@ -0,0 +1,514 @@
...
+subrouter.use("/:siteName", settingsSubrouter)
+
+const mockGenericAxios = mockAxios.create()
+mockGenericAxios.patch.mockResolvedValue({
+  data: [],
+})
+// const migrateSpy = jest
+//   .spyOn(ReviewsRouter, "checkIfSiteIsUnmigrated")
+//   .mockResolvedValue(true)
+
```
Comment (seaerchin) [inline src/integration/Privatisation.spec.ts:219]: pls remove

File: src/integration/Privatisation.spec.ts [hunk 6]
```diff
@@ -0,0 +1,514 @@
...
+        },
+      })
+      if (!actualDeployment) fail()
+      const { encryptionIv, encryptedPassword } = actualDeployment
+      expect(encryptionIv).toEqual(null)
+      expect(encryptedPassword).toEqual(null)
+      expect(mockGenericAxios.patch).not.toHaveBeenCalled()
+    })
+
+    it("should successfully set a password for public amplify sites", async () => {
```
Comment (seaerchin) [inline src/integration/Privatisation.spec.ts:438]: will it cause the sites to become private? if so, how are we tsting this?

File: src/integration/Privatisation.spec.ts [hunk 7]
```diff
@@ -0,0 +1,514 @@
...
+      const actual = await request(app).get(
+        `/${MOCK_REPO_NAME_TWO}/repo-password`
+      )
+
+      // Assert
+      expect(actual.statusCode).toEqual(200)
+      expect(actual.body).toMatchObject(expected)
+    })
+
+    it("should successfully return empty string for an amplify site with no password", async () => {
```
Comment (seaerchin) [inline src/integration/Privatisation.spec.ts:287]: can we have another test case where we want to remove but we also provide a password?

Comment (alexanderleegs) [reply src/integration/Privatisation.spec.ts:244]: We actually already have this - it's the above test for invalid IV, I've updated the name of the test to make it clearer! https://github.com/isomerpages/isomercms-backend/pull/807/commits/79f7fce7969a0c816a0893ee2babad6b253e5fa3

File: src/integration/Privatisation.spec.ts (hunk 6 above)
Comment (alexanderleegs) [reply src/integration/Privatisation.spec.ts:394]: I'm checking this via the `mockAxios.patch` call, which only occurs for this route when the github repo is changed (to make private or public)

File: src/integration/Privatisation.spec.ts (hunk 5 above)
Comment (alexanderleegs) [reply src/integration/Privatisation.spec.ts]: https://github.com/isomerpages/isomercms-backend/pull/807/commits/273d92cfa7f5630eb7868085c2089bae503ce597

File: src/integration/Privatisation.spec.ts (hunk 4 above)
Comment (alexanderleegs) [reply src/integration/Privatisation.spec.ts]: https://github.com/isomerpages/isomercms-backend/pull/807/commits/dff24365c7db806004989049a0faeba56a1b11d3

File: src/integration/Privatisation.spec.ts (hunk 3 above)
Comment (alexanderleegs) [reply src/integration/Privatisation.spec.ts]: dff24365

File: src/routes/v2/authenticatedSites/__tests__/Settings.spec.js (hunk 2 above)
Comment (alexanderleegs) [reply src/routes/v2/authenticatedSites/__tests__/Settings.spec.js]: f30f538a

File: src/utils/__tests__/crypto-utils.spec.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    const fakeIv = "12345678901234561234567890123456"
+    try {
+      decryptPassword(ENCRYPTED_PASSWORD, fakeIv)
+    } catch (e) {
+      return
+    }
+    fail()
+  })
+
+  it("should not match if different password is given", async () => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/utils/__tests__/crypto-utils.spec.ts]: You're right that it's the same, i've removed it! As for the diff iv test case, we actually already have this - it's the above test for invalid IV, I've updated the name of the test to make it clearer! https://github.com/isomerpages/isomercms-backend/commit/79f7fce7969a0c816a0893ee2babad6b253e5fa3

File: src/utils/__tests__/crypto-utils.spec.ts [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+  })
+
+  it("should have invalid IV throw error", async () => {
+    const fakeIv = "12345678901234561234567890123456"
+    try {
+      decryptPassword(ENCRYPTED_PASSWORD, fakeIv)
+    } catch (e) {
+      return
+    }
+    fail()
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/utils/__tests__/crypto-utils.spec.ts]: Changed in https://github.com/isomerpages/isomercms-backend/commit/79f7fce7969a0c816a0893ee2babad6b253e5fa3!

File: src/validators/RequestSchema.js [hunk 10]
```diff
@@ -276,8 +276,8 @@ const UpdateSettingsRequestSchema = Joi.object().keys({
 })
 
 const UpdateRepoPasswordRequestSchema = Joi.object().keys({
-  password: Joi.string(),
-  enablePassword: Joi.boolean(),
+  password: Joi.string().required().allow(""),
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/validators/RequestSchema.js]: Yeah that's a good point, i've tightened the request schema in https://github.com/isomerpages/isomercms-backend/pull/807/commits/56b33edc0dcf49ae5b374b9a092c58a197702942

File: src/services/db/__tests__/GitHubService.spec.ts [hunk 11]
```diff
@@ -912,4 +913,39 @@ describe("Github Service", () => {
...
+        refEndpoint,
+        { private: shouldBePrivate },
+        {
+          headers,
+        }
+      )
+      expect(resp.isOk()).toEqual(true)
+    })
+
+    it("should modify the repo privacy accordingly if changing to false", async () => {
```
Comment (kishore03109) [inline src/services/db/__tests__/GitHubService.spec.ts:937]: [nit]: "should modify the repo privacy accordingly if changing to false" can be "should be able to make repo public"

**Review by kishore03109 — APPROVED** [review]

File: src/services/db/__tests__/GitHubService.spec.ts (hunk 11 above)
Comment (alexanderleegs) [reply src/services/db/__tests__/GitHubService.spec.ts]: https://github.com/isomerpages/isomercms-backend/pull/807/commits/6154dfd83020b5a74d045fe4d8b7dc94fadf997e

File: src/integration/Privatisation.spec.ts [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+      if (!actualDeployment) fail()
+      const { encryptionIv, encryptedPassword } = actualDeployment
+      expect(encryptionIv).toEqual(null)
+      expect(encryptedPassword).toEqual(null)
+      expect(mockGenericAxios.patch).not.toHaveBeenCalled()
+    })
+
+    it("should successfully set a password for public amplify sites", async () => {
+      // Arrange
+      const newPassword = "blah"
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/integration/Privatisation.spec.ts]: Yeah that's a good point - [6e5f766](https://github.com/isomerpages/isomercms-backend/pull/807/commits/6e5f766e80c89f557959a94753b843378b3af05e) and [49ab4cd](https://github.com/isomerpages/isomercms-backend/pull/807/commits/49ab4cda090b57685002502de30fdcd97d9fd96b)
